### PR TITLE
Error for negative start index in string section_issues#4871

### PR DESCRIPTION
**Pull request by AnshMehta1**
This pull request aims to fix issue #4871. Basically , a compile-time check is done for negative start index in case of character array splicing or substring section and correspondingly , a semantic error is thrown for negative start index. Changes in codebase are made accordingly to check for character array and then for start index.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -6801,6 +6801,24 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
             var_name = handle_templated(x.m_func, ASR::is_a<ASR::Template_t>(*ASRUtils::get_asr_owner(owner_sym)),
                 x.m_temp_args, x.n_temp_args, x.base.base.loc);
         }
+
+        if (x.m_args != nullptr && x.m_args->m_start != nullptr) {
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:6805]:
```suggestion
        if (x.m_args != nullptr && x.m_args[0]->m_start != nullptr) {
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -6801,6 +6801,24 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+
+            if((x.m_args->m_start->type) == 4){
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:6807]: `x.m_args` is a list, so you need to access accordingly

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:6805]: You cannot do something like this, `x.m_args` is a list of `fnarg_t`. See the definition of node:

```c
    struct FuncCallOrArray_t // Constructor
    {
        const static exprType class_type = exprType::FuncCallOrArray;
        typedef expr_t parent_type;
        expr_t base;
        char* m_func;
        struct_member_t* m_member; size_t n_member; // Sequence
        fnarg_t* m_args; size_t n_args; // Sequence
        keyword_t* m_keywords; size_t n_keywords; // Sequence
        fnarg_t* m_subargs; size_t n_subargs; // Sequence
        decl_attribute_t** m_temp_args; size_t n_temp_args; // Sequence
    };
```

The comment `// Sequence` means it is a list containing elements of a kind. Thereby you need to iterate over it.

For accessing first element:

```cpp
AST::fnarg_t arg_0 = x.m_args[0]
```

Now you can perform whatever you wish

Comment (AnshMehta1) [reply src/lfortran/semantics/ast_common_visitor.h:6805]: I made these changes locally but now I am facing Problem in Allocate . Basically , Allocate allows negative starting index but my changes are throwing semantic error for allocate part. I am not able to resolve this.

Comment (AnshMehta1) [reply src/lfortran/semantics/ast_common_visitor.h]: program arrays_39
implicit none


real, allocatable :: array(:)
integer :: ndim

ndim = 10

allocate(array(-(ndim/2):ndim))

print *, lbound(array > 0, 1)
call f(array > 0)

contains

    subroutine f(array1)
        logical, intent(in) :: array1(:)
        print *, lbound(array1, 1), ubound(array1, 1)
        if( lbound(array1, 1) /= 1 ) error stop
        if( ubound(array1, 1) /= 16 ) error stop
    end subroutine

end program arrays_39

**Review by Pranavchiku — COMMENTED** [review]
```diff
diff --git a/src/lfortran/semantics/ast_common_visitor.h b/src/lfortran/semantics/ast_common_visitor.h
index 9e510a93e..eec5f9d66 100644
--- a/src/lfortran/semantics/ast_common_visitor.h
+++ b/src/lfortran/semantics/ast_common_visitor.h
@@ -3790,6 +3790,7 @@ public:
     ASR::asr_t* create_ArrayRef(const Location &loc, AST::fnarg_t* m_args,
         size_t n_args, AST::fnarg_t* m_subargs, size_t n_subargs,
         ASR::expr_t* v_expr, ASR::symbol_t *v, ASR::symbol_t *f2) {
+        std::cout<<"create_ArrayRef\n";
         ASR::ttype_t* root_v_type = ASRUtils::type_get_past_pointer(
             ASRUtils::symbol_type(v));
         size_t n_dims = ASRUtils::extract_n_dims_from_ttype(root_v_type);
@@ -4028,6 +4029,7 @@ public:
                 this->visit_expr(*(m_subargs[0].m_step));
                 step = ASRUtils::EXPR(tmp);
                 step = CastingUtil::perform_casting(step, int_type, al, loc);
+                std::cout<<"here2"<<std::endl;
                 return ASR::make_StringSection_t(al, loc, array_item, l,
                         r, ASRUtils::EXPR(tmp), char_type, arr_ref_val);
             } else {
@@ -4085,7 +4087,7 @@ public:
                         char_type = ASRUtils::TYPE(ASR::make_Character_t(al, loc,
                                         1, a_len, a_len_expr, ASR::string_physical_typeType::PointerString));
                     }
-
+                    std::cout<<"here3"<<std::endl;
                     return ASR::make_StringSection_t(al, loc, v_Var, l,
                             r, casted_step, char_type, arr_ref_val);
                 }
@@ -6903,6 +6905,9 @@ public:
                 throw SemanticError("Subroutine `" + var_name + "` called as a function ", x.base.base.loc);
             }
         }
+        if ( ASRUtils::is_array(ASRUtils::symbol_type(v)) && ASRUtils::is_character(*ASRUtils::symbol_type(v))) {
+            std::cout<<"Character array"<<std::endl;
+        }
         if (!is_common_variable
             && ( ASR::is_a<ASR::Variable_t>(*v) || is_external_procedure )
             && (!ASRUtils::is_array(ASRUtils::symbol_type(v)))
```

**Comment by AnshMehta1** [comment]
After registering my test in tests.toml , my some ci tests are failing , so what might be the possibl reason?

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -6912,6 +6912,19 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 throw SemanticError("Subroutine `" + var_name + "` called as a function ", x.base.base.loc);
             }
         }
+        else{
+
+            if((v->type == 9) && (ASRUtils::is_character(*ASRUtils::symbol_type(v)))){
+            
+                if (x.m_args != nullptr && x.m_args[0].m_start != nullptr) {
+
+                    if((x.m_args[0].m_start->type) == 4){
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:6921]: Use an enum, and `is_a<>` idiom.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 4]
```diff
@@ -6912,6 +6912,19 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                        throw SemanticError("Substring `start` is negative", x.base.base.loc);
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:6922]:
```suggestion
                        throw SemanticError("The first index in string section is negative", x.base.base.loc);
```

**Pull request closed without merging**